lib/backend/k8s/resources/customnoderesource.go

```diff
@@ -0,0 +1,436 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resources
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	"github.com/projectcalico/libcalico-go/lib/errors"
+
+	log "github.com/Sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
+
+// Action strings - used for context logging.
+type action string
+
+const (
+	actApply  action = "Apply"
+	actCreate action = "Create"
+	actUpdate action = "Update"
+)
+
+// K8sResourceConverter defines an interface to map between the model and the
+// annotation representation of a resource.,
+type CustomK8sNodeResourceConverter interface {
+	// ListInterfaceToNodeAndName converts the ListInterface to the node name
+	// and resource name.
+	ListInterfaceToNodeAndName(model.ListInterface) (string, string, error)
+
+	// KeyToNodeAndName converts the Key to the node name and resource name.
+	KeyToNodeAndName(model.Key) (string, string, error)
+
+	// NodeAndNameToKey converts the Node name and resource name to a Key.
+	NodeAndNameToKey(string, string) (model.Key, error)
+}
+
+// customK8sNodeResourceClient implements the K8sResourceClient interface and provides a generic
```

Comment:
wrong capitalization in the comment here (and `NewCustomK8sNodeResourceClientConfig`)